### PR TITLE
PSBTv2

### PR DESCRIPTION
**Pull request by scgbckbone**

### CODE REVIEW

File: psbt_dump.py [hunk 1]
```diff
@@ -241,13 +257,33 @@ def dump(psbt, hex_output, bin_output, testnet, base64, show_addrs):
...
             ADP = testnetwork.address if testnet else network.address
 
+            if section == "globals" and key[0] == PSBT_GLOBAL_TX_MODIFIABLE:
+                print("\tbitfield: %s\n" % bin(val[0]))
+                if val[0] & 0b1:
+                    print("\tInputs can be modified.")
+                if val[0] & 0b10:
+                    print("\tOutputs can be modified.")
+                if val[0] & 0b100:
+                    print("\tHas SIGHASH_SINGLE flag.")
```
Comment (doc-hex) [inline psbt_dump.py:284]: display illegal values here (unknown bits set => show hex)

File: psbt_dump.py [hunk 2]
```diff
@@ -241,13 +257,33 @@ def dump(psbt, hex_output, bin_output, testnet, base64, show_addrs):
                     print("'%s' = 0x%x = %d = %s\n" % (b2a_hex(val), nn, nn, SIGHASH_MAP[nn]))
                 else:
                     print("'%s' = 0x%x = %d\n" % (b2a_hex(val), nn, nn))
+
+                if key[0] == PSBT_GLOBAL_VERSION:
+                    print("\tPSBTv%d\n" % nn)
```
Comment (doc-hex) [inline psbt_dump.py:262]: track version of current file, show errors when incorrect fields vs. version are used

File: psbt_dump.py [hunk 3]
```diff
@@ -276,6 +312,16 @@ def dump(psbt, hex_output, bin_output, testnet, base64, show_addrs):
                     print("(unable to parse txn)")
                     raise
 
+            if section == "outputs" and key[0] == PSBT_OUT_AMOUNT:
+                sats = struct.unpack("q", val)[0]
+                print('\t%d sats\n' % sats)
```
Comment (doc-hex) [inline psbt_dump.py:317]: show BTC amt too

File: psbt_dump.py [hunk 4]
```diff
@@ -22,33 +21,50 @@
...
+PSBT_IN_WITNESS_UTXO 	         = 1
+PSBT_IN_PARTIAL_SIG 	         = 2
+PSBT_IN_SIGHASH_TYPE 	         = 3
+PSBT_IN_REDEEM_SCRIPT 	         = 4
+PSBT_IN_WITNESS_SCRIPT 	         = 5
+PSBT_IN_BIP32_DERIVATION 	     = 6
+PSBT_IN_FINAL_SCRIPTSIG 	     = 7
+PSBT_IN_FINAL_SCRIPTWITNESS      = 8
+PSBT_IN_POR_COMMITMENT           = 9   # Proof of Reserves
+PSBT_IN_RIPEMD160                = 10  # 0x0a
```
Comment (doc-hex) [inline psbt_dump.py:44]: use hex

File: psbt_dump.py [hunk 5]
```diff
@@ -22,33 +21,50 @@
...
+PSBT_IN_OUTPUT_INDEX             = 15  # 0x0f
+PSBT_IN_SEQUENCE                 = 16  # 0x10
+PSBT_IN_REQUIRED_TIME_LOCKTIME   = 17  # 0x11
+PSBT_IN_REQUIRED_HEIGHT_LOCKTIME = 18  # 0x12
+PSBT_IN_TAP_KEY_SIG              = 19  # 0x13
+PSBT_IN_TAP_SCRIPT_SIG           = 20  # 0x14
+PSBT_IN_TAP_LEAF_SCRIPT          = 21  # 0x15
+PSBT_IN_TAP_BIP32_DERIVATION     = 22  # 0x16
+PSBT_IN_TAP_INTERNAL_KEY         = 23  # 0x17
+PSBT_IN_TAP_MERKLE_ROOT          = 24  # 0x18
```
Comment (doc-hex) [inline psbt_dump.py:58]: BIP numbers were useful but got lost

File: psbt_dump.py [hunk 6]
```diff
@@ -416,6 +498,17 @@ def dump(psbt, hex_output, bin_output, testnet, base64, show_addrs):
                         # XXX decode implied p2sh addresses here too?
                 except:
                     print("(unable to parse POSSIBLE multisig redeem script)")
+    print()
+    if version:
+        if version == 0 and v2_fields:
+            print("INVALID: PSBT v0 requires exclusion of PSBT v2 fields.")
```
Comment (doc-hex) [inline psbt_dump.py:504]: unreachable

Comment (scgbckbone) [reply psbt_dump.py:504]: changed `if version:` to `if version is not None`

what do you mean by unreachable ?

**Pull request merged**